### PR TITLE
Intrusive list implementation and tests

### PR DESCRIPTION
**Pull request by forGGe**
Inspired by common practice in C-kernels and boost::intrusive
containers.

Provided version of a list is suitable for standard library algorithms
and range-for constructs. It also guarantees safety when deleting nodes
during iteration.

Intended to use in composing fashion.

### CODE REVIEW

File: lib/containers/export/ecl/list.hpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+    //! Destroys iterator
+    ~list_iter();
+
+    //! Proceed to the next item in a list.
+    list_iter& operator ++();
+    //! Postfix version of the operatoris.
+    list_iter operator ++(int);
+
+    //! Moves back to the previous item in a list.
```
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Missing, must be removed

File: lib/containers/export/ecl/list.hpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    list_iter& operator --();
+    //! Postfix version of the operator.
```
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Missing, must be removed

File: lib/containers/export/ecl/list.hpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+inline typename std::enable_if< std::is_literal_type< T2 >::value, size_t >::type
+constexpr offset_of(T1 T2::*member) {
+    constexpr T2 object {};
+    return size_t(&(object.*member)) - size_t(&object);
+}
+
+/// \todo move this to right module
+template< typename T1, typename T2 >
+inline size_t constexpr offset_of(T1 T2::*member) {
+    return reinterpret_cast< size_t >(&(((T2*)0)->*member));
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: Why not `nullptr`?

File: lib/containers/export/ecl/list.hpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+// Taken from https://gist.github.com/graphitemaster/494f21190bb2c63c5516
+/// \todo move this to right module
+template< typename T1, typename T2 >
+inline typename std::enable_if< std::is_literal_type< T2 >::value, size_t >::type
+constexpr offset_of(T1 T2::*member) {
+    constexpr T2 object {};
+    return size_t(&(object.*member)) - size_t(&object);
+}
+
+/// \todo move this to right module
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: You use two doxygen styles in the file: `///` and `//!`.

File: lib/containers/export/ecl/list.hpp [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+private:
+    list_node *m_next; //!< Next node in a list. Equals to this if the list is empty.
+    list_node *m_prev; //!< Previous node in a list. Equals to this if the list is empty.
+};
+
+
+//------------------------------------------------------------------------------
+
+
+list_node::list_node()
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: Implementing the class in the header file is not ok unless these are inline methods (they are not).

File: lib/containers/export/ecl/list.hpp [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    //! Adds a node before this node.
+    //! \param[in] Node to add in the list
+    void add_before(list_node &node);
+
+    //! Unlinks this node from a list.
+    //! \details Has no effect if list is empty.
+    void unlink();
+
+    //! Checks if node is connected to any other node.
+    bool linked() const;
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: The better name is `is_linked()`.

File: lib/containers/export/ecl/list.hpp [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+    //! Returns a pointer to a next list node.
+    list_node* next() const;
+
+    //! Returns a pointer to a previous list node.
+    list_node* prev() const;
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: What's your code style? Align pointer star left or right?

File: lib/containers/export/ecl/list.hpp [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    //! Returns iterator to end of the list,
+    //! provided that this node is a head.
+    //! \warning Dereferencing end iterator is illegal
+    list_iter< T, Mptr > end();
+
+    //! Checks if list is empty.
+    bool empty() const;
+
+    //! Push element to the end of a list.
+    void push_back(T &t);
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: No `push_front()` or `pop()`?

File: lib/containers/export/ecl/list.hpp [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+    //! Acessess an object.
+    T& operator *();
+
+    //! Compares two iterators.
+    //! \warning Comparing iterators from different lists is illegal.
+    //! \return True if iterators point to the same object within the same list.
+    bool operator ==(const list_iter& other);
+
+    //! \copydoc opeator==().
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: Not sure `\copydoc` is ok, cause return values are different.

File: lib/containers/export/ecl/list.hpp [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    bool operator !=(const list_iter< T, Mptr >& other);
+
+private:
+    list_node *m_cur; //! Current list node.
+    list_node *m_tmp; //! Temporary list node. Used to when deleting nodes.
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: Should be `Used for deleting nodes.`

File: lib/containers/export/ecl/list.hpp [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    // Advance pointer. The current node has been possibly deleted.
+    m_cur = m_tmp;
+    m_tmp = m_tmp->next();
+    return *this;
+}
+template< typename T, list_node T::* Mptr >
+list_iter< T, Mptr > list_iter< T, Mptr >::operator ++(int)
+{
+    list_iter tmp{*this};
+    operator++();
```
Comment (rasendubi) [inline lib/containers/export/ecl/list.hpp]: Why not `++*this;`?

File: lib/containers/export/ecl/list.hpp (hunk 3 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Fixed

File: lib/containers/export/ecl/list.hpp (hunk 4 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Fixed

File: lib/containers/export/ecl/list.hpp (hunk 5 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Fixed

File: lib/containers/export/ecl/list.hpp (hunk 6 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: TBD in #41

File: lib/containers/export/ecl/list.hpp (hunk 7 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: TBD in #41 
For now, I place asterisk to the right unless in a type of a return value.

File: lib/containers/export/ecl/list.hpp (hunk 8 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Non yet needed.

File: lib/containers/export/ecl/list.hpp (hunk 9 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Fixed.

File: lib/containers/export/ecl/list.hpp (hunk 10 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Fixed

File: lib/containers/export/ecl/list.hpp (hunk 11 above)
Comment (forGGe) [inline lib/containers/export/ecl/list.hpp]: Looks minimalistic. Fixed.

**Comment by forGGe** [comment]
Close #39

**Pull request merged**